### ISSUE TITLE
[kvdb-rocksdb] Give Database::iter a concrete type

### ISSUE THREAD

**Issue by maciejhirsz** [paritytech/parity-common#437]
The [return type on `Database::iter` in `kvdb-rocksdb` is `impl Iterator<...>`](https://github.com/paritytech/parity-common/blob/master/kvdb-rocksdb/src/lib.rs#L569), which limits the usability of the iterator for downstream users, in particular we can't put it into [an associated type for a trait](https://github.com/paritytech/substrate-archive/pull/108#discussion_r514096408) and have to use a boxed iterator trait object instead.

It would be nice to have a public type alias for the concrete type, or a newtype wrapper around it to hide internals if necessary.

**Comment by ordian** [comment]
Thanks for bringing this up.

I'm however skeptical we should do this:
1. It probably has zero impact on performance. `iter` is dominated by I/O, unless you call it in a loop and don't actually iterate, which is a very contrived scenario. I'd like to avoid breaking changes for no gain.
2. I'm not sure it's possible to implement as a associated type, because the iterator type has a reference, so it seems to be blocked on GAT in Rust: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bbf27237552047ffc86df9eee475c6e3.

I'm happy to be proven wrong on both points.

**Comment by maciejhirsz** [comment]
It's doable with a trait impl on a ref with lifetime set, though that requires adding another trait and some pretty verbose extra trait bounds, which wouldn't be super pretty, so fair enough.

**Comment by ordian** [comment]
closing as wontfix since we're deprecating kvdb-rocksdb
